pkg/ingester/metrics.go

```diff
@@ -0,0 +1,211 @@
+			Help: "The current number of series in memory.",
+		}),
+		memUsers: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_ingester_memory_users",
+			Help: "The current number of users in memory.",
+		}),
+		memSeriesCreatedTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingester_memory_series_created_total",
+			Help: "The total number of series that were created per user.",
+		}, []string{"user"}),
+		memSeriesRemovedTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingester_memory_series_removed_total",
+			Help: "The total number of series that were removed per user.",
+		}, []string{"user"}),
+	}
+
+	if r != nil {
+		r.MustRegister(
+			m.flushQueueLength,
+			m.ingestedSamples,
+			m.ingestedSamplesFail,
+			m.queries,
+			m.queriedSamples,
+			m.queriedSeries,
+			m.queriedChunks,
+			m.memSeries,
+			m.memUsers,
+			m.memSeriesCreatedTotal,
+			m.memSeriesRemovedTotal,
+		)
+	}
+
+	return m
+}
+
+// TSDB shipper metrics. We aggregate metrics from individual TSDB shippers into
+// a single set of counters, which are exposed as Cortex metrics.
+type shipperMetrics struct {
+	dirSyncs        *prometheus.Desc // sum(thanos_shipper_dir_syncs_total)
+	dirSyncFailures *prometheus.Desc // sum(thanos_shipper_dir_sync_failures_total)
+	uploads         *prometheus.Desc // sum(thanos_shipper_uploads_total)
+	uploadFailures  *prometheus.Desc // sum(thanos_shipper_upload_failures_total)
+
+	regsMu sync.RWMutex                    // custom mutex for shipper registry, to avoid blocking main user state mutex on collection
+	regs   map[string]*prometheus.Registry // One prometheus registry (used by shipper) per tenant
+}
+
+func newShipperMetrics(r prometheus.Registerer) *shipperMetrics {
+	m := &shipperMetrics{
+		regs: make(map[string]*prometheus.Registry),
+
+		dirSyncs: prometheus.NewDesc(
+			"cortex_ingester_shipper_dir_syncs_total",
+			"TSDB: Total dir sync attempts",
+			nil, nil),
+		dirSyncFailures: prometheus.NewDesc(
+			"cortex_ingester_shipper_dir_sync_failures_total",
+			"TSDB: Total number of failed dir syncs",
+			nil, nil),
+		uploads: prometheus.NewDesc(
+			"cortex_ingester_shipper_uploads_total",
+			"TSDB: Total object upload attempts",
+			nil, nil),
+		uploadFailures: prometheus.NewDesc(
+			"cortex_ingester_shipper_upload_failures_total",
+			"TSDB: Total number of failed object uploads",
+			nil, nil),
+	}
+
+	if r != nil {
+		r.MustRegister(m)
+	}
+	return m
+}
+
+func (sm *shipperMetrics) Describe(out chan<- *prometheus.Desc) {
+	out <- sm.dirSyncs
+	out <- sm.dirSyncFailures
+	out <- sm.uploads
+	out <- sm.uploadFailures
+}
+
+func (sm *shipperMetrics) Collect(out chan<- prometheus.Metric) {
+	gathered := make(map[string][]*dto.MetricFamily)
+
+	regs := sm.shipperRegistries()
+	for userID, r := range regs {
+		m, err := r.Gather()
+		if err != nil {
+			level.Debug(util.Logger).Log("msg", "failed to gather metrics from TSDB shipper", "user", userID, "err", err)
```

Comment:
I would `Warn` instead of `Debug`. This shouldn't happen and - if happens - we get unreliable metrics.

Reply:
I've changed it. Although user cannot really do anything about it, at least we will see it in logs easier. Hopefully it never happens.

Reply:
> at least we will see it in logs easier

Exactly, that's the point. If it happens it's likely to be a bug. Noticing would allow the user to submit us an issue.